staging/src/k8s.io/client-go/discovery/cached/disk/compactable_disk_cache.go

```diff
@@ -0,0 +1,234 @@
+// cache can breach this threshold. When cache compaction occurs, it will resize
+// the disk cache below this "cacheSize".
+//
+// Trigger:
+//
+//	disk cache compactions are triggered when a file is accessed and the
+//	last compaction occurred over a day ago. Every cache file access checks
+//	when the last compaction occurred by checking the modification time
+//	on an empty file in the cache -- "lastcompactiontime". File access
+//	time for a cached file is not used directly because it is not well
+//	supported by all OSes.
+//
+// Concurrency:
+//
+//	Within the same process: the goroutines of a single process are
+//	coordinated using "sync/singleflight" functionality.
+//
+//	Among different processes: if (for example) a separate kubectl command
+//	compacts the disk cache at the same time, separate processes can compete
+//	to delete the same files. This code must be tolerant to attempting to
+//	delete files already deleted.
+type CompactableDiskCache struct {
+	cacheDir  string
+	cacheSize int64
+	delegate  httpcache.Cache
+}
+
+// Get returns the []byte representation of a cached response and a bool
+// set to true if the value isn't empty. This function can trigger a
+// disk cache compaction which runs in a separate goroutine. Touches
+// cached file to update file modification to match access time.
+func (cdc *CompactableDiskCache) Get(key string) ([]byte, bool) {
+	// If the file exists, update file modification time = access time.
+	_ = touchFile(sanitize(key))
+	go possiblyCompactDiskCache(cdc.cacheDir, cdc.cacheSize)
+	return cdc.delegate.Get(key)
+}
+
+// Set stores the []byte representation of a response against a key.
+func (cdc *CompactableDiskCache) Set(key string, responseBytes []byte) {
+	cdc.delegate.Set(key, responseBytes)
+}
+
+// Delete removes the value associated with the key.
+func (cdc *CompactableDiskCache) Delete(key string) {
+	cdc.delegate.Delete(key)
+}
+
+// NewCompactableDiskCache returns a newly created CompactableDiskCache,
+// which includes a newly created sumDiskCache delegate.
+func NewCompactableDiskCache(cacheDir string, cacheSize int64) *CompactableDiskCache {
+	return &CompactableDiskCache{
+		cacheDir:  cacheDir,
+		cacheSize: cacheSize,
+		delegate: &sumDiskCache{
+			disk: diskv.New(diskv.Options{
+				PathPerm: os.FileMode(0750),
+				FilePerm: os.FileMode(0660),
+				BasePath: cacheDir,
+				TempDir:  filepath.Join(cacheDir, ".diskv-temp"),
+			}),
+		},
+	}
+}
+
+// possiblyCompactDiskCache checks if the disk cache compaction
+// trigger condition has occurred, and if it has, it runs (only
+// one) compaction.
+func possiblyCompactDiskCache(cacheDir string, cacheSize int64) {
+	compactionFile := filepath.Join(cacheDir, lastCompactionFilename)
+	if shouldCompactCache(compactionFile) {
+		klog.V(7).Infoln("Disk cache compaction triggered")
+		// "singleflight" ensures that only one disk compaction occurs at a time within
+		// the current process.
+		var requestGroup singleflight.Group
```

Comment:
I would expect that the Group needs to be global or a member variable? This seems to construct a new one each time if I am reading it correctly.